src/garage/tf/algos/batch_polopt2.py

```diff
@@ -0,0 +1,281 @@
+
+        """
+        baselines = []
+        returns = []
+
+        max_path_length = self.max_path_length
+
+        if self.flatten_input:
+            paths = [
+                dict(
+                    observations=(self.env_spec.observation_space.flatten_n(
+                        path['observations'])),
+                    actions=(
+                        self.env_spec.action_space.flatten_n(  # noqa: E126
+                            path['actions'])),
+                    rewards=path['rewards'],
+                    env_infos=path['env_infos'],
+                    agent_infos=path['agent_infos']) for path in paths
+            ]
+        else:
+            paths = [
+                dict(
+                    observations=path['observations'],
+                    actions=(
+                        self.env_spec.action_space.flatten_n(  # noqa: E126
+                            path['actions'])),
+                    rewards=path['rewards'],
+                    env_infos=path['env_infos'],
+                    agent_infos=path['agent_infos']) for path in paths
+            ]
+
+        if hasattr(self.baseline, 'predict_n'):
+            all_path_baselines = self.baseline.predict_n(paths)
+        else:
+            all_path_baselines = [
+                self.baseline.predict(path) for path in paths
+            ]
+
+        for idx, path in enumerate(paths):
+            path_baselines = np.append(all_path_baselines[idx], 0)
+            deltas = (path['rewards'] + self.discount * path_baselines[1:] -
+                      path_baselines[:-1])
+            path['advantages'] = np_tensor_utils.discount_cumsum(
+                deltas, self.discount * self.gae_lambda)
+            path['deltas'] = deltas
+
+        for idx, path in enumerate(paths):
+            # baselines
+            path['baselines'] = all_path_baselines[idx]
+            baselines.append(path['baselines'])
+
+            # returns
+            path['returns'] = np_tensor_utils.discount_cumsum(
+                path['rewards'], self.discount)
+            returns.append(path['returns'])
+
+        obs = np.concatenate([path['observations'] for path in paths])
+        actions = np.concatenate([path['actions'] for path in paths])
+        rewards = np.concatenate([path['rewards'] for path in paths])
+        returns = np.concatenate(returns)
+        baselines = np.concatenate(baselines)
+
+        agent_infos_path = [path['agent_infos'] for path in paths]
+        agent_infos = dict()
+        for key in self.policy.state_info_keys:
+            agent_infos[key] = np.concatenate(
+                [infos[key] for infos in agent_infos_path])
+
+        env_infos = [path['env_infos'] for path in paths]
+        env_infos = tensor_utils.stack_tensor_dict_list([
+            tensor_utils.pad_tensor_dict(p, max_path_length) for p in env_infos
+        ])
+
+        valids = [np.ones_like(path['returns']) for path in paths]
```

Comment:
Shouldn't this be a numpy array? This is a list

Reply:
Same for lengths

Reply:
valids is not returned so I guess it’s fine and yes, lengths should be a numpy array

Reply:
are you sure we shouldn't return valids? why not, if we have it?

Reply:
We will be passing `lengths`, which serves the same purpose as `valids`. But sure, we can also return `valids` as another format of `lengths`. For example, we have `valids` to be 
[1, 1, 1, 1, 0, 0, 0, 0, 0, 0] and we will have `lengths` to be [4].

Reply:
i think it might be convenient in the future for some users, and converting from `lengths` is a pain. Can you make sure it's a `numpy.ndarray`?

We can reconsider this when you and KR looks at the policy+sampler APIs. For now I think it's best to include it.

Reply:
Sounds good